File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = -1;
    
    void helper(TreeNode *root, int &k)
    {
        if (!root or k == 0) return;
        
        helper(root->left, k);
        k--;
        if (k == 0) ans = root->val;
        helper(root->right, k);
    }
    
    int kthSmallest(TreeNode* root, int k) {
        int K = k;
        helper(root, K);
        return ans;
    }
};

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        
        ans = -1
        
        def inorder(node):
            if not node:
                return
            
            nonlocal k
            
            if k >= 0:
                inorder(node.left)
                k -= 1
                nonlocal ans
                if k == 0: ans = node.val
                # print(ans, k)
                inorder(node.right)
        
        inorder(root)
        return ans
